### PR TITLE
Feature/5973/retain state of percentage fee

### LINKED ISSUE #5973: [Order Creation] Retain state of "Percentage fee" checkbox when editing a fee
Discussed in p1646250815730879-slack-C02KUCFCSFP. 

https://user-images.githubusercontent.com/198826/156442515-95c94ae1-1509-44b3-ac2f-32f45a9b53f0.mov

## Steps

1. Tapped on Fee to add a fee.
2. Checked Percentage fee.
3. Entered 10 (percentage)
4. Pressed Done.
5. Tapped on Fee to edit that fee.
6. The Percentage fee checkbox is not checked. It's showing me the Amount instead of the percentage that I entered.

## Discussion

This seems like a limitation since we save the fixed amount even if they entered a percentage. There's no easy way to tell if they previously entered a percentage. Core also behaves similarly. 

We could probably use `meta` to keep the state.

### PR DESCRIPTION
**Pull request by atorresveiga**
Closes: #5973

### Description
This PR aims to retain the fee state by using a local fee manager as the source of truth. We were facing the problem that the app calculated the percentage and created an amount fee with the result when selecting a percentage fee. We don't manage such a thing as a percentage fee on the server-side, so, by using it as the source of truth, we couldn't be sure, because all fees were saved as amount fees (FeeLine), whether the user was creating a percentage fee or an amount fee. 
Another problem I found was that when we updated the total order amount by adding new products, the percentage fee, as was calculated once, didn't update. 
What should happen with the percentage fee if we delete all products? 🤔
One solution proposed by Nick was to update the design to reflect what the app was doing (calculating the percentage but stored as an amount) and be consistent with what we have in code. The design needed to update, though.
This PR tries a different solution. It uses a Local Fee Manager to save the Fee state locally (in the device) and use it as the source of truth for Fees. It also solves the other problems mentioned before (recalculate percentages fees when the total order updates and removes the percentage fee when we don't have products). This solution is possible because we don't save the order draft to resume the creation in another device.
I needed to make some concessions and calculate the percentage based on product total and not order total because order total includes taxes. When auto-updated, the percentages ended up with an infinite loop (percentages depend on total and affect taxes -> taxes updates total) if I use order total. Maybe this could be improved to total = product total + shipping total. 🤔
This PR will solve the state problem of the percentage fee and be consistent with the design intentions. Nevertheless, I needed to change the total used for calculating the fees, and it will require some context to maintain this code in the future (the source of truth is the Fee Manager unless the Fee is removed). I would love a confidence check if it is something to push forward or is better to change the UI and be consistent with what we can do on the BE.

### Testing instructions
1. Orders -> New Order -> Create Order
2. Add Product -> Select Product -> Check product total updates.
3. Add Fee -> Enable **Percentage fee** Switch
4. Enter 10 in Percentage(%) EditText -> Press Done -> Check Fee is 10% of Products Total
5. Press + to add another Product -> Check Fee is and Products Total updates (Check fee is 10% of Products Total)
6. Remove all products -> Check the Fee is also removed
7. Press Add Fee -> Check  Enable **Percentage fee** Switch is not visible
8. Enter 100 in Amount ($) EditText -> Press Done -> Check Fee is $100

### Images/gif

Result using products total
https://user-images.githubusercontent.com/18119390/160474169-56ecb5d7-b392-4b07-8976-f0aafe25a8c1.mp4

Result using order total
https://user-images.githubusercontent.com/18119390/160475155-bae00b91-2fa6-4d5b-b388-cc27246f3ed1.mp4

### CODE REVIEW

**Comment by hichamboushaba** [comment]
I tested the PR, and I see two big issues, I don't know how hard they would be to fix:

1. We make multiple requests after each update to the order (I noticed 4 requests specifically, i didn't spend time to understand what's causing it):

https://user-images.githubusercontent.com/1657201/160615405-0e73f188-ae7d-4b56-b76f-5cbc06deb68e.mov

2. If the Activity gets recreated after, we lose all the entered fees, since they are not saved to the `SavedState`.

And I have one remark regarding the change itself (a subjective one 🙂), from an architectural perspective, I think that having two sources of truths will just make the code hard to follow, here we have both the `_orderDraft` and the fees are coming from `LocalFeeManager`, which means more work to keep things in sync, and understand who the data flows.
I think an easier-to-follow approach would to refactor how we model the data itself, instead of using plain `Order` in `_orderDraft`, we can probably think of a model that can represents fees in a more clear way that can represent percentage fees, and do the mapping only when calling the API 🤔.

One last point, I didn't follow all the discussions around the subject, but do we feel like this feature justifies the added complexity? And also this will have a different user experience than wp-admin, especially when it comes to taxes.

**Comment by shiki** [comment]
Just highlighting Joe's suggestion/question in https://github.com/woocommerce/woocommerce-android/issues/5973#issuecomment-1081351873 that may be simpler:

> Until we can overcome the technical issue, the only thing I can think of is adding some clarification.
>
> - Change “Percentage fee” to “Calculate as a percentage”
> - Add a caption to the field: “Calculated amount: $23.99”
>
> You’ll enter the percentage, see the dollar value, and confirm your decision. Then on the order details screen, you’ll recognise the dollar value when you see it again.
> 
> Is that possible?

**Comment by atorresveiga** [comment]
>We make multiple requests after each update to the order (I noticed 4 requests specifically, i didn't spend time to understand what's causing it):

Ups! Sorry for that. I was recording the video to show why I chose to make the percentage calculation based on the product's total instead of the order's total and uploaded the change by mistake while fixing detekt warnings. 
The reason it makes all those requests is because of a cycle dependency. If we set a percentage fee and update the number of products, as shown in the video above, with all the requests, the order's total gets affected (we included a new product), so taxes and the total of the order will increase. As the percentage fee depends on the total of the order, it will increase again once the order is refreshed, also affecting the taxes and the total of the order, and this will iterate several times. Because of this, I made the percentage calculations based on the products' total, avoiding the cycle dependency (percentage -> total, total -> percentage).

>If the Activity gets recreated after, we lose all the entered fees, since they are not saved to the SavedState.

Good catch! I updated the code to make use of SavedState.

>I think an easier-to-follow approach would to refactor how we model the data itself, instead of using plain Order in _orderDraft, we can probably think of a model that can represents fees in a more clear way that can represent percentage fees, and do the mapping only when calling the API 🤔.

Great idea! Wrapping the Order model in a new model that knows how to manage all fees will make the logic easy to follow. Still, it will have some complexity because it needs to know when to sync, update the order's total, refresh percentages, etc, and will have a different user experience than wp-admin. Joe's suggestion would be a simpler approach and we can iterate on this solution later.
I will close this PR and create another one with Joe's suggestion.

**Comment by atorresveiga** [comment]
I will close this PR and create another one following Joe's suggestion of:
Change “Percentage fee” to “Calculate as a percentage”
Add a caption to the field: “Calculated amount: $23.99”

**Comment by hichamboushaba** [comment]
Thanks so much @atorresveiga for the work on this.

> Ups! Sorry for that. I was recording the video to show why I chose to make the percentage calculation based on the product's total instead of the order's total and uploaded the change by mistake while fixing detekt warnings.

Yep, I missed the fact that there are two recordings in the description 🤦, sorry 🙂.
